### PR TITLE
Simplified Spark implementation, Added Proto support to Spark.

### PR DESCRIPTION
**Pull request by blublinsky**
### How was this patch tested?

Comment. This works correctly locally, but for running on the server we need to replace protobuf jars, packaged with Spark:

    assemblyShadeRules in assembly := Seq(
      ShadeRule.rename("com.google.protobuf.**" -> "shadeproto.@1").inAll
    )
To make it local, the following is required:

      dependencyOverrides += "com.fasterxml.jackson.core" % "jackson-core" % "2.11.2" ,
      dependencyOverrides += "com.fasterxml.jackson.core" % "jackson-databind" % "2.11.2",
      dependencyOverrides += "com.fasterxml.jackson.module" % "jackson-module-scala_2.12" % "2.11.2",

      dependencyOverrides += "com.google.protobuf" % "protobuf-java" % "3.11.4",

We need to figure out how to do the same for running on cluster. https://hadoopsters.com/2019/05/08/how-to-override-a-spark-dependency-in-client-or-cluster-mode/ seems to provide an option

### CODE REVIEW

**Comment by RayRoestenburg** [comment]
Please also update the examples @blublinsky ? I'm assuming implicits will be needed there for encoders

File: core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala [hunk 1]
```diff
@@ -58,12 +54,7 @@ class SparkStreamletContextImpl(
 
     val rawDataset = src.select($"value").as[Array[Byte]]
 
-    val dataframe: Dataset[Row] = rawDataset.mapPartitions { iter ⇒
-      val avroDecoder = new SparkAvroDecoder[In](schema)
-      iter.map(avroDecoder.decode)
-    }(inRowEncoder)
-
-    dataframe.as[In]
+    rawDataset.map(inPort.codec.decode(_))
```
Comment (maasg) [inline core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:57]: @blublinsky Don't you get a `SerializationError` if you do this? Did you try on a cluster?
Back then, I had many issues with `avro.Schema` because it's not serializable. 
I would expect that applying a transformation where the codec instance is is passed will attempt to serialize the `Schema` member of it.

Comment (maasg) [reply core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:57]: It would be great if it works. It makes it follow the same pattern that we are using everywhere else.

Comment (blublinsky) [reply core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:57]: Yes it does. Just finished testing it on the cluster. The key here is that Avro generated class is just a simple case class, for which Spark creates a default encoder, that is used here.

**Comment by blublinsky** [comment]
So on of the keys was to rebuild the base image to update protobuf version and add Sparksql-scalapb. Once this was done (part of PR), everything works on the server as well

File: core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala (hunk 1 above)
Comment (maasg) [reply core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:57]: Update: bummer... 
I don't think it's going to work. 
I created a serialization test as follows:
```scala
class SerializationSpec extends WordSpec with SparkScalaTestSupport with TryValues {
  "Avro Codec" should {
    "be serializable" in {
      val avroInlet = AvroInlet[Data]("name")
      val deser     = avroInlet.codec.decode _
      try {
        SparkEnv.get.closureSerializer.newInstance().serialize(deser)
        // pass!
      } catch {
        case ex: Throwable => fail(ex.toString)
      }
    }
"Sanity check" should {
    "not be serializable" in {
      val localResource   = new Thread()
      val notSerializable = () => localResource
      try {
        SparkEnv.get.closureSerializer.newInstance().serialize(notSerializable)
        fail("Thread should not be serializable")
      } catch {
        case ex: Throwable => () // pass
      }

    }
  }
  }
```
and it fails with the same Exception that I was getting back in the day.
```
[info] - should be serializable *** FAILED ***
[info]   java.io.NotSerializableException: org.apache.avro.Schema$RecordSchema
[info]   Serialization stack:
[info]          - object not serializable (class: org.apache.avro.Schema$RecordSchema, value: {"type":"record","name":"Data","namespace":"cloudflow.spark.avro","fields":[{"name":"id","type":"int"},{"name":"name","type":"string"}]})
[info]          - field (class: cloudflow.streamlets.avro.AvroCodec, name: avroSchema, type: class org.apache.avro.Schema)
[info]          - object (class cloudflow.streamlets.avro.AvroCodec, cloudflow.streamlets.avro.AvroCodec@64f4c593)
[info]          - element of array (index: 0)
[info]          - array (class [Ljava.lang.Object;, size 1)
[info]          - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
[info]          - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class cloudflow.spark.SerializationSpec, functionalInterfaceMethod=scala/Function1.apply:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic cloudflow/spark/SerializationSpec.$anonfun$new$3:(Lcloudflow/streamlets/avro/AvroCodec;[B)Lcloudflow/spark/avro/Data;, instantiatedMethodType=([B)Lcloudflow/spark/avro/Data;, numCaptured=1])
[info]          - writeReplace data (class: java.lang.invoke.SerializedLambda)
[info]          - object (class cloudflow.spark.SerializationSpec$$Lambda$12337/821591673, cloudflow.spark.SerializationSpec$$Lambda$12337/821591673@629795ce) (SerializationSpec.scala:33)
[info] Sanity check
[info] - should not be serializable
```

Comment (blublinsky) [reply core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:57]: Of course it will not. You do not have encoder. Take a look at my tests https://github.com/lightbend/Spark-sensors-proto. All works as a champ

File: examples/spark-sensors-proto/README.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+```
+
+* Undeploy.
+
+```bash
+$ kubectl cloudflow undeploy spark-sensors
+```
+### Protobuf vs Avro
+
+This project demonstarte usage of protobuf's for message encoding. Protobufs are enabled through usage of
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:186]:
```suggestion
This project demonstrates the usage of Protobuf for message encoding. Protobuf is enabled through usage of
```

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc [hunk 3]
```diff
@@ -28,6 +28,26 @@ codec of the data they manipulate.
 Currently, Cloudflow supports Avro and hence the classes are named `AvroInlet` and `AvroOutlet`.
 Each outlet also allows the user to define a partitioning function that will be used to partition the data.
 
+== Usage of different encodings in Spark streamlet
+
+Currently Cloudflow supports two different data encoding types - Avro and Protobuf. Both can be used for Spark streamlet,
+but you should should remember several rules. Protobuf support is based on https://scalapb.github.io/docs/sparksql/[Scala-PB Spark],
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:34]:
```suggestion
but you should remember several rules. Protobuf support is based on https://scalapb.github.io/docs/sparksql/[Scala-PB Spark],
```

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc [hunk 4]
```diff
@@ -28,6 +28,26 @@ codec of the data they manipulate.
...
+which comes with its own `implicits` - `scalapb.spark.Implicits._` which is not compatible with `cloudflow.spark.sql.SQLImplicits._`.
+When using these `implicits` the following rules have to be observed:
+
+* If protobuf's encoding is not used, always use `cloudflow.spark.sql.SQLImplicits._`.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:38]: Can't we use `org.apache.spark.sql.SQLImplicits` instead?

Comment (RayRoestenburg) [reply docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:38]: (You might have told me before but I can't remember why we still have these)

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc [hunk 5]
```diff
@@ -28,6 +28,26 @@ codec of the data they manipulate.
...
+* If you are using at least one protobuf encoding, you need to use `scalapb.spark.Implicits._`.
+Note that this `implicits` do not include default implicit supporting Scala classes. So if in addition to
+protobuf, you are using Avro and or any custom scala types in your Spark processing, you have to add implicit encoders, for these
+types. For example, if you are using the following type in your calculations:
+----
+case class Rate(timestamp: Timestamp, value: Long)
+----
+In your code you shoud have a definition of the corresponding encoder:
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:46]:
```suggestion
In your code you should have a definition of the corresponding encoder:
```

File: examples/spark-sensors-proto/README.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
+## Spark-based Cloudflow Application using protobuf encoding
+
+### Problem Definition
+
+In this application we use a an ingress to generate random data from a set of virtual sensors, use Spark to emit moving average values for each sensor id and the report the values to the console egress.
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:5]:
```suggestion
In this application we use an ingress to generate random data from a set of virtual sensors, use Spark to emit moving average values for each sensor id and the report the values to the console egress.
```

File: examples/spark-sensors-proto/README.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+All Protobuf message definitions can be found [here](src/main/protobuf).
+Protobuf encoding is enabled through usage of `ProtoInlet` and `Protoutlet` and usage of the `scalapb.spark.Implicits._` import
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:7]:
```suggestion
Protobuf encoding is enabled through usage of `ProtoInlet` and `Protoutlet` and usage of the `scalapb.spark.Implicits._` import.
```

File: examples/spark-sensors-proto/README.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+Note: `scalapb.spark.Implicits._` import is incompatible with `cloudflow.spark.sql.SQLImplicits._`.
+The two can be used together in the same class. As a result, if an implementation is using any of the custom
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:10]:
```suggestion
The two cannot be used together in the same class. As a result, if an implementation is using any of the custom
```

File: examples/spark-sensors-proto/README.md [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+scala case classes (or Avro) encoders for these classes have to be defined. For example a
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:11]:
```suggestion
scala case classes (or Avro), encoders for these classes have to be defined. For example a
```

File: examples/spark-sensors-proto/README.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+$ kubectl cloudflow undeploy spark-sensors
+```
+### Protobuf vs Avro
+
+This project demonstrates the usage of Protobuf for message encoding. Protobuf is enabled through usage of
+[Scala-PB Spark](https://scalapb.github.io/docs/sparksql/).
+ScalaPB comes with its own `implicits` - `scalapb.spark.Implicits._` which is not compatible with `cloudflow.spark.sql.SQLImplicits._`
+The following are the rules for using imports.
+
+* When protobufs are not used, you should use `cloudflow.spark.sql.SQLImplicits._` in all classes as required - see example
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:240]: see example?

File: examples/spark-sensors-proto/README.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+* When protobufs are used, you should use `scalapb.spark.Implicits._`. Here are the caveats. 
+You may want to import `cloudflow.spark.sql.SQLImplicits.StringToColumn` to convert $"col name" into a Column.
+If you use custom scala classes in you Spark calculations, you need to explicitely create encoder for it. See
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:243]:
```suggestion
If you use custom scala classes in you Spark calculations, you need to explicitly create encoder for it. See
```

File: examples/spark-sensors-proto/README.md [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+[SparkRandomGenDataIngress](src/main/scala/sensors/proto/SparkRandomGenDataIngress.scala) for example
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:244]:
```suggestion
[SparkRandomGenDataIngress](src/main/scala/sensors/proto/SparkRandomGenDataIngress.scala) for example.
```

File: examples/spark-sensors-proto/README.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+* Usage of both imports in the same class will lead to a problem
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:245]:
```suggestion
* Usage of both imports in the same class will lead to problems.
```

File: examples/spark-sensors-proto/README.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+Also note dependency overwrites necessary to use the right version of `com.fasterxml.jackson.core` versions
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/README.md:247]: Why is this necessary?

File: examples/spark-sensors-proto/project/cloudflow-plugins.sbt [hunk 15]
```diff
@@ -0,0 +1,5 @@
+// Resolver for the cloudflow-sbt plugin
+//
+resolvers += Resolver.url("cloudflow", url("https://lightbend.bintray.com/cloudflow"))(Resolver.ivyStylePatterns)
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/project/cloudflow-plugins.sbt:3]: This is not necessary, sbt-cloudflow is published to sonatype.

File: examples/spark-sensors-proto/src/main/scala/sensors/proto/MovingAverageSparklet.scala [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sensors.proto
+
+import cloudflow.spark.sql.SQLImplicits.StringToColumn
```
Comment (RayRoestenburg) [inline examples/spark-sensors-proto/src/main/scala/sensors/proto/MovingAverageSparklet.scala:19]: Can you explain what this does?

File: examples/spark-sensors/README.md [hunk 17]
```diff
@@ -1,33 +1,72 @@
-## Spark-based Cloudflow Application
+## Spark-based Cloudflow Application using Avro encoding
 
 ### Problem Definition
 
 In this application we use a an ingress to generate random data from a set of virtual sensors, use Spark to emit moving average values for each sensor id and the report the values to the console egress.
+All Avro message definitions can be found [here](src/main/avro).
+Avro encoding is enabled through usage of `AvroInlet` and `AvroOutlet`and usage of the `cloudflow.spark.sql.SQLImplicits._` import
```
Comment (RayRoestenburg) [inline examples/spark-sensors/README.md:7]:
```suggestion
Avro encoding is enabled through usage of `AvroInlet` and `AvroOutlet`and usage of the `cloudflow.spark.sql.SQLImplicits._` import.
```

File: external/multi-base-images/spark/buildLightbendSpark.sh [hunk 18]
```diff
@@ -45,11 +47,18 @@ $DIR/spark/dev/make-distribution.sh --name cloudflow-2.12 --tgz -Pscala-2.12 -Ph
 
 # build the Spark image
 tar -zxvf spark-$SPARK_VERSION-bin-cloudflow-2.12.tgz
-cd $DIR/spark/spark-$SPARK_VERSION-bin-cloudflow-2.12
+# replace protobuf jar to support sparksql-scalapb and load Scala-PB
```
Comment (RayRoestenburg) [inline external/multi-base-images/spark/buildLightbendSpark.sh:50]: Do you have an idea for getting this to work without needing our own build of Spark?

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc (hunk 4 above)
Comment (blublinsky) [reply docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:38]: cloudflow.spark.sql.SQLImplicits and org.apache.spark.sql.SQLImplicits are identical. org.apache.spark.sql.SQLImplicits is used in conjunction with Spark session, see https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html 
Gerard created cloudflow.spark.sql.SQLImplicits so that it can be ised anywhere in the code.

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc (hunk 3 above)
Comment (blublinsky) [reply docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:34]: Fixed

File: docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc (hunk 5 above)
Comment (blublinsky) [reply docs/shared-content-source/docs/modules/develop/pages/use-spark-streamlets.adoc:46]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 6 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:5]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 7 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:7]: fixed

File: examples/spark-sensors-proto/README.md (hunk 8 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:10]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 9 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:11]: fixed

File: examples/spark-sensors-proto/README.md (hunk 10 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:240]: fixed

File: examples/spark-sensors-proto/README.md (hunk 11 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:243]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 12 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:244]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 13 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:245]: Fixed

File: examples/spark-sensors-proto/README.md (hunk 14 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/README.md:247]: Was there before. Removed, based on changes to cloudflow here

File: examples/spark-sensors-proto/project/cloudflow-plugins.sbt (hunk 15 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/project/cloudflow-plugins.sbt:3]: Fixed

File: examples/spark-sensors-proto/src/main/scala/sensors/proto/MovingAverageSparklet.scala (hunk 16 above)
Comment (blublinsky) [reply examples/spark-sensors-proto/src/main/scala/sensors/proto/MovingAverageSparklet.scala:19]: It Converts $"col name" into a `org.apache.spark.sql.Column`. See this line
        .groupBy(window($"ts", "1 minute", "30 seconds"), $"src", $"gauge")

File: examples/spark-sensors/README.md (hunk 17 above)
Comment (blublinsky) [reply examples/spark-sensors/README.md:7]: Fixed

File: external/multi-base-images/spark/buildLightbendSpark.sh (hunk 18 above)
Comment (blublinsky) [reply external/multi-base-images/spark/buildLightbendSpark.sh:50]: Not really, unless Spark will update their protobuf version.
At any rate we are using out own image so that we can inject spark-entrypoint.sh
Its not really a different build its just replacing protobufs jar and adding scalapb jar

File: external/multi-base-images/spark/buildLightbendSpark.sh [hunk 19]
```diff
@@ -45,11 +47,18 @@ $DIR/spark/dev/make-distribution.sh --name cloudflow-2.12 --tgz -Pscala-2.12 -Ph
...
+rm spark-$SPARK_VERSION-bin-cloudflow-2.12/jars/protobuf-java-*.*
```
Comment (RayRoestenburg) [inline external/multi-base-images/spark/buildLightbendSpark.sh:51]: This script does not work for me when I just run it, it stops with:
```
VERSION=
```
executing parts of it in my shell does work. I'm ok with that now, it might have always been that way, or it is something in my shell. This should all be added to our normal build anyway, so we can fix it then.

File: core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala (hunk 1 above)
Comment (RayRoestenburg) [reply core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala:56]: Verified that everything works on the cluster.
